dependency/health_service_test.go

```diff
@@ -675,6 +691,136 @@ func TestHealthServiceQuery_Fetch(t *testing.T) {
 	}
 }
 
+// TestHealthServiceQuery_Fetch_SamenessGroup tests fetching services with sameness group
+// and asserts that failover works when a service in the sameness group fails and consul-template
+// reacts to the change.
+func TestHealthServiceQuery_Fetch_SamenessGroup(t *testing.T) {
+	// Arrange - set up test data
+	catalog := testClients.Consul().Catalog()
+
+	partitionOne := "sg-partition-1"
+	partitionTwo := "sg-partition-2"
+	nodeOne := "node" + partitionOne
+	nodeTwo := "node" + partitionTwo
+	samenessGroup := "test-sameness-group"
+
+	require.NoError(t, testClients.createConsulPartition(partitionOne))
+	require.NoError(t, testClients.createConsulPartition(partitionTwo))
+	require.NoError(t, testClients.createConsulSamenessGroups(samenessGroup, partitionOne, partitionTwo))
+
+	// Register services in foo partition as critical so they will fall back to default partition service
+	svcName := "sameness-group-service"
+	registerSvc := func(service, node, partition, status string) {
+		checkName := fmt.Sprintf("%s:%s", service, node)
+		svcRegistration := &api.CatalogRegistration{
+			Service: &api.AgentService{
+				ID:        service,
+				Service:   service,
+				Port:      12345,
+				Partition: partition,
+				Namespace: "default",
+				Connect:   &api.AgentServiceConnect{},
+			},
+			Partition: partition,
+			Node:      node,
+			Address:   "127.0.0.1",
+			Checks: api.HealthChecks{
+				&api.HealthCheck{
+					Node:        node,
+					CheckID:     checkName,
+					Name:        checkName,
+					Status:      status,
+					ServiceID:   svcName,
+					ServiceName: svcName,
+				},
+			},
+		}
+
+		_, err := catalog.Register(svcRegistration, nil)
+		if err != nil {
+			t.Fatal(err)
+		}
+	}
+
+	// set up service in each partition
+	registerSvc(svcName, nodeOne, partitionOne, "passing")
+	registerSvc(svcName, nodeTwo, partitionTwo, "passing")
+
+	// Act - fetch the service
+	query := fmt.Sprintf("%s?sameness-group=%s&partition=%s", svcName, samenessGroup, partitionOne)
+	d, err := NewHealthServiceQuery(query)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	svcs, meta, err := d.Fetch(testClients, nil)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	healthServices := svcs.([]*HealthService)
```

Comment:
```suggestion
        // Assert that the service instance returned is from the first listed partition.
        // This is expected because the service is healthy so it shouldn't failover
        // to other partitions.
	healthServices := svcs.([]*HealthService)
```